src/storage/examples/upsert_open_loop.rs

```diff
@@ -0,0 +1,1253 @@
+    ) -> Self {
+        // bounded??
+        let (tx, rx): (
+            _,
+            crossbeam_channel::Receiver<(Vec<Batch>, Sender<Result<Batch<KVAndPrevious>, Error>>)>,
+        ) = crossbeam_channel::unbounded();
+
+        let instance_path = temp_dir
+            .join(format!("worker_id:{}", worker_id.to_string()))
+            .join(format!("source_id:{}", source_id.to_string()));
+
+        if destroy_before_use {
+            DB::destroy(&rocksdb::Options::default(), &*instance_path).unwrap();
+        }
+
+        let mut rocks_options = rocksdb::Options::default();
+        rocks_options.create_if_missing(true);
+
+        if use_vector_memtable {
+            rocks_options.set_memtable_factory(rocksdb::MemtableFactory::Vector);
+            // Required to use the vector memtable.
+            rocks_options.set_allow_concurrent_memtable_write(false);
+        }
+
+        if let Some(global_env) = global_env {
+            rocks_options.set_env(&global_env);
+        }
+
+        let db: DB = DB::open(&rocks_options, instance_path).unwrap();
+        std::thread::spawn(move || {
+            let mut wo = rocksdb::WriteOptions::new();
+            wo.disable_wal(!use_wal);
+
+            'batch: while let Ok((batches, resp)) = rx.recv() {
+                let size: usize = batches.iter().map(|b| b.len()).sum::<usize>();
+
+                // TODO(guswynn): this should probably be lifted into the upsert operator.
+                if size == 0 {
+                    let _ = resp.send(Ok(Vec::new()));
+                    continue;
+                }
+
+                let gets = db.multi_get(
+                    batches
+                        .iter()
+                        .flat_map(|b| b.iter().map(|(k, _)| k.as_slice())),
+                );
+
+                let mut previous = Vec::new();
+                let mut writes = rocksdb::WriteBatch::default();
+
+                // TODO(guswynn): sort by key before writing.
+                for ((k, v), get) in batches.into_iter().flat_map(|b| b.into_iter()).zip(gets) {
+                    writes.put(k.as_slice(), v.as_slice());
+
+                    match get {
+                        Ok(prev) => {
+                            previous.push((k, v, prev));
+                        }
+                        Err(e) => {
+                            // Give up on the batch on errors.
+                            let _ = resp.send(Err(e));
+                            continue 'batch;
+                        }
+                    }
+                }
+                match db.write_opt(writes, &wo) {
+                    Ok(()) => {}
+                    Err(e) => {
+                        // Give up on the batch on errors.
+                        let _ = resp.send(Err(e));
+                        continue 'batch;
+                    }
+                }
+                debug!("finished writing batch size({size}) for worker {worker_id}, source {source_id}");
```

Comment:
I don't know how hard it is to hook into the metrics, but RocksDB has a toooon of coverage and it might be worth surfacing some things like mem/block cache/bloom filter hit rates, size on disk vs size of dataset